x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/GCMPacketsCipherInputStream.java

```diff
@@ -0,0 +1,455 @@
+ *
+ * This is obviously NOT thread-safe.
+ */
+public class GCMPacketsCipherInputStream extends FilterInputStream {
+
+    static class GCMPacketsWithMarkCipherInputStream extends GCMPacketsCipherInputStream {
+
+        private GCMPacketsWithMarkCipherInputStream(InputStream in, SecretKey secretKey, int mode, long packetIndex, int nonce,
+                                                    Provider provider) {
+            super(in, secretKey, mode, packetIndex, nonce, provider);
+        }
+
+        private ByteArrayOutputStream markWriteBuffer = null;
+        private ByteArrayInputStream markReadBuffer = new ByteArrayInputStream(new byte[0]);
+        private boolean markTriggeredForCurrentPacket = false;
+        private long markPacketIndex;
+        private int markReadLimit;
+
+        @Override
+        public int read() throws IOException {
+            ensureOpen();
+            // in case this is a reseted stream that has buffered part of the ciphertext
+            if (markReadBuffer.available() > 0) {
+                return markReadBuffer.read();
+            }
+            int cipherByte = super.read();
+            // if buffering of the ciphertext is required
+            if (markTriggeredForCurrentPacket && cipherByte != -1) {
+                markWriteBuffer.write(cipherByte);
+            }
+            return cipherByte;
+        }
+
+        @Override
+        public int read(byte b[], int off, int len) throws IOException {
+            ensureOpen();
+            // in case this is a reseted stream that has buffered part of the ciphertext
+            int bytesReadFromMarkBuffer = markReadBuffer.read(b, off, len);
+            if (bytesReadFromMarkBuffer != -1) {
+                return bytesReadFromMarkBuffer;
+            }
+            int cipherBytesCount = super.read(b, off, len);
+            // if buffering of the ciphertext is required
+            if (markTriggeredForCurrentPacket && cipherBytesCount > 0) {
+                markWriteBuffer.write(b, off, cipherBytesCount );
+            }
+            return cipherBytesCount;
+        }
+
+        @Override
+        public long skip(long n) throws IOException {
+            ensureOpen();
+            if (markReadBuffer.available() > 0) {
+                return markReadBuffer.skip(n);
+            }
+            int bytesAvailable = super.available();
+            bytesAvailable = Math.min(bytesAvailable, Math.toIntExact(n));
+            if (markTriggeredForCurrentPacket) {
+                byte[] temp = new byte[bytesAvailable];
+                int bytesRead = super.read(temp);
+                markWriteBuffer.write(temp);
+                return bytesRead;
+            } else {
+                return super.skip(n);
+            }
+        }
+
+        @Override
+        public int available() throws IOException {
+            ensureOpen();
+            return markReadBuffer.available() + super.available();
```

Comment:
Well, not really. From a high level theoretical pov returning only `markReadBuffer.available()` might seem correct. But in practice, most of the time, `markReadBuffer.available()` would be exactly `0` because buffering only happens in limited cases when mark is not called on packet boundary and reset has not yet exhausted the bytes until the next packet buffer.

Reply:
Ok :) I don't think it matters either way as we're not making use of `available` anywhere in the first place.